copy and move constructor for data sources which respects the const correctness of the data type

SimeonEhrig opened the pull request:
We allow to copy/move data sources with the following inner data types to other objects with the same type with the corresponding (different) inner data type
  - mutable -> mutable
  - const -> const
  - mutable -> const

# ToDo
- [x] merge #270 
- [x] implement constructors for `alpaka::SharedBuffer`
- [x] verify that move of `alpaka::SharedBuffer` is working and no copy is done
- [x] implement constructors for `alpaka::View`
- [x] finish `alpaka::SharedBuffer` tests
- [x] implement bool operator for `alpaka::mdspan` and `alpaka::View` and extent `alpaka::concepts::impl::IMdSpan`

SimeonEhrig committed:
implement copy and move constructor for alpaka::mdspan which respects the const correctness of the data type

- We allow to copy/move data sources with the following inner data types to other objects with the same type with the corresponding (different) inner data type
  - mutable -> mutable
  - const -> const
  - mutable -> const

- Only alpaka::mdspan is implemented at the moment. alpaka::sharedBuffer and alpaka::View is missing.
- Only compile tests yet. Didn't check if the `alpaka::mdspan` is moved instead copied.

SimeonEhrig committed:
fixed leftovers from PR #270

mehmetyusufoglu commented on tests/unit/mem/constCorrectness.cpp:
```diff
@@ -31,93 +33,171 @@ void requiresMutableBufferMd(concepts::MdSpan auto buffer)
+    concepts::Vector auto pitches = alpaka::calculatePitchesFromExtents<int>(extents);
+
+    MdSpan mdspan{ptr, extents, pitches};
+    func_copy_by_value<int, int&>(mdspan);
```
camelCase function names?

SimeonEhrig replied:
Changed in commit 71fb3da59913e8e70a782fa8800d4292202ba007

SimeonEhrig committed:
implement copy and move constructor for alpaka::sharedBuffer which respects the const correctness of the data type

- Constructing a `alpaka::SharedBuffer<int const>` from a `alpaka::SharedBuffer<int>` does not work yet, because `alpaka::View` does not support the construction with different element types. `alpaka::SharedBuffer` inherited from `alpaka::View`.
- Create concept `InnerTypeAllowedCast`, which checks if inner raw element type is the same and if the conversion from a const/mutable element type to another const/mutable element type is allowed.
- Add lifetime test to `alpaka::SharedBuffer` to verify that the buffer is correct copied or moved.

SimeonEhrig committed:
fix for header check

SimeonEhrig committed:
implement copy and move constructor for alpaka::View which respects the const correctness of the data type

SimeonEhrig committed:
fixed copy assignment operator of sharedBuffer

- now all copy and move tests of sharedBuffer are working

SimeonEhrig committed:
add function to check if memory access of a MdSpan and View is valid (e.g. after a move)

SimeonEhrig committed:
add function to check if memory access of a MdSpan and View is valid (e.g. after a move)
